irma-core/state-machine.js

```diff
@@ -11,70 +11,105 @@ module.exports = class StateMachine {
     this._disabledTransitions = [];
   }
 
+  /* Deprecated
   currentState() {
     return this._state;
-  }
+  } *
 
+  /* Deprecated
   isEndState() {
     return this._inEndState;
-  }
+  } */
 
+  /* Deprecated
   isValidTransition(transition) {
     if (this._inEndState || this._disabledTransitions.includes(transition))
       return false;
     return transitions[this._state][transition] != undefined;
-  }
+  } */
 
   addStateChangeListener(func) {
     this._listeners.push(func);
   }
 
+  // TODO: Add docs
   transition(transition, payload) {
-    this._performTransition(transition, false, payload)
+    return this._performTransition({transition, payload});
   }
 
+  // TODO: Add docs
   finalTransition(transition, payload) {
-    this._performTransition(transition, true, payload);
+    return this._performTransition({transition, payload, isFinal: true});
   }
```

Comment:
Same holds for these functions. I'd prefer to use the same construction here as `_performTransition` has, but this is again an incompatible API change for custom plugins. So again, do we want to give developers time to make this change or can we change this right away?

Reply:
So when exactly should these be used in this new version?

Reply:
If you want to do a transition without actually checking whether it is possible. When it is not possible, the Promise is rejected, so you can detect in that way that it fails. I can imagine this being useful when doing an `abort` or `fail`, which are possible from almost any state. These are a little bit less cumbersome to use.

The distinction between `transition` and `finalTransition` is maybe a bit weird now, because in `selectTransition` there is simply an option `isFinal` to determine this. To straighten the API I thought it would be nice to also combine these two functions into one accordingly.

It is a bit of a code smell indeed, because it would be better to always use `selectTransition`. So you would be in favour of deprecating these onces too if I understand you right?

Reply:
I think I would prefer that. 
Alternatively, if you think `abort` and `fail` will be used most often we could two shortcuts for just these two:
```
Abort() => this._performTransition({'Aborted', true})
Fail(err) => this._performTransitioin({'Failed', true, err})
```
And all other transactions should be performed using `selectTransaction`.

Reply:
I deprecated them, but I kept them in as aliases for now with a deprecation warning. I chose to also show the warning when debugging is `false` as an extra motivation for people to stop using it.